src/TorchSharp/Dataset.cs

```diff
@@ -49,8 +64,31 @@ public void Dispose()
 
                     protected virtual void Dispose(bool disposing)
                     {
+                        IDataset<Dictionary<string, string>> a = null;
+                        IDataset<IReadOnlyDictionary<string, string>> b = a;
                     }
                 }
+
+                /// <summary>
+                /// The base interface for all Datasets.
+                /// </summary>
+                public interface IDataset<out T> : IDisposable
+                {
+                    /// <summary>
+                    /// Size of dataset
+                    /// </summary>
+                    long Count { get; }
+
+                    /// <summary>
+                    /// Get tensor according to index
+                    /// </summary>
+                    /// <param name="index">Index for tensor</param>
+                    /// <returns>Tensors of index. DataLoader will catenate these tensors into batches.</returns>
+                    [IndexerName("DatasetItems")]
+                    T this[long index] { get; }
```

Comment:
It's using indexers instead of `GetTensor` in the new interface. I'm not really sure about its `IndexerName`.

And... shall we remove `GetTensor` at the same time? Currently `Dataset<T>` is keeping a abstract `GetTensor` to ensure compatibility.